settings: Allow markdown in custom profile fields description.

roanster007 opened the pull request:
This commit allows markdown in description of custom profile fields in the label and hint fields.

Fixes #28612

<!-- Describe your pull request here.-->


<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<img width="566" alt="image" src="https://github.com/zulip/zulip/assets/97049524/2c171b08-6eca-4996-be31-99448c9011fb">
<br>
<img width="380" alt="image" src="https://github.com/zulip/zulip/assets/97049524/c777f213-6ec4-4eef-bf19-58822b33ce7a">
<br>
<img width="389" alt="image" src="https://github.com/zulip/zulip/assets/97049524/e159bc2c-378c-48ea-9420-3689c7353b8c">
<br>
<img width="292" alt="image" src="https://github.com/zulip/zulip/assets/97049524/e53e78af-7c83-45a2-bb03-2a34817e008b">
<br>
<img width="198" alt="image" src="https://github.com/zulip/zulip/assets/97049524/9b4079eb-ac0e-42c2-a5b0-18c11113819d">
<br>
<img width="215" alt="image" src="https://github.com/zulip/zulip/assets/97049524/68344749-e131-49bf-b522-7f7a92a2e2ab">


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

roanster007 committed:
settings: Allow markdown in custom profile fields description.

This commit allows markdown in description of custom profile
fields in the label and hint fields.

Fixes #28612

alya commented:
Thanks! I think this'll be a cool feature.

alya commented:
I noticed that the Markdown isn't showing up properly in SETTINGS / PROFILE.

 
<img width="407" alt="Screenshot 2024-03-14 at 2 56 20 PM" src="https://github.com/zulip/zulip/assets/2090066/b28438ae-7b4d-4cef-be75-2128d5738226">

alya commented:
For a mention:

<img width="416" alt="Screenshot 2024-03-14 at 2 58 02 PM" src="https://github.com/zulip/zulip/assets/2090066/5a1ccf39-419b-4022-b73c-8526f06d64ce">

alya commented:
It looks good in the other places I tested! (custom profile settings, profile modal, user card)

alya commented:
Can we do better with how field names are displayed in the user card tooltip?

<img width="311" alt="Screenshot 2024-03-14 at 2 56 51 PM" src="https://github.com/zulip/zulip/assets/2090066/847c9dc8-1d1a-4bf9-8d35-e2943ffb5d07">


It looks good in the other places I tested! (custom profile settings, profile modal)

roanster007 committed:
user_card: Update the user card tippy content for custom fields.

Previously, the tippy content for custom fields used to be the field
name. However, since this field name could now markdown,
the tippy content would also include the rectangular brackets and
parenthesis of the hyperlinks, and mentions.

Hence, this commit adds the tippy_content to profile_field, which
contains only the text field of the hyperlink, removing the links,
and only the mentioned user's name.

roanster007 commented:
Also, if there's a mention in the field name, then the tooltip on the user card would be `@**Zoe**`.  Is it ok, or do you want the `*` to be removed?

alya commented:
What changes did you make to the tooltips, if any?

roanster007 commented:
So for the tooltips on user card of custom fields -- if the field is a mention, the tooltip would be `@Zoe` rather than `@**Zoe**` and if the field is a url, then tooltip would be `google` rather than `[google](google.com)`

alya commented:
OK, that seems reasonable to me, thanks!

You can ping @sahil839 for a review once conflicts are resolved.

roanster007 commented:
@sahil839 could you have a look at this!

sahil839 commented on zerver/openapi/zulip.yaml:
```diff
@@ -14997,6 +15031,11 @@ paths:
+                              type: string
+                              description: |
+                                The markdown renderedname of the external account
+                                provider
```
You would need to add "Changes" entry here. You can search for "Changes" in the file to see how it is done.

sahil839 commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,11 @@ format used by the Zulip server that they are interacting with.
 
 ## Changes in Zulip 9.0
 
+**Feature level 246**
```
You would also need to update the `API_FEATURE_LEVEL` field in `version.py`.

roanster007 replied:
thanks for letting me know! updated

sahil839 commented on zerver/models/custom_profile_fields.py:
```diff
@@ -69,7 +69,9 @@ class CustomProfileField(models.Model):
     name = models.CharField(max_length=NAME_MAX_LENGTH)
+    rendered_name = models.TextField(null=True, default=None)
     hint = models.CharField(max_length=HINT_MAX_LENGTH, default="")
+    rendered_hint = models.TextField(null=True, default=None)
```
Can we just set these fields for existing `CustomProfileField` objects as well? We did so when adding `rendered_description` field for streams in 73d26c8b28.

roanster007 replied:
I updated it to use `default=""` for existing objects. However, while fetching the initial state, I am popping the `rendered_name` and `rendered_hint` fields if `rendered_name == ""` (indicating for existing objects). Do you think this is the right way to do it, or do I handle it at the frontend?

sahil839 commented on web/src/user_profile.js:
```diff
@@ -269,6 +269,15 @@ function render_manage_profile_content(user) {
 
+function get_tippy_content(field_name) {
+    const hyperlink = /\[([^[\]]*?)]\((.*?)\)/g;
+    const mentions = /\*{1,2}([^*]+)\*{1,2}/g;
```
Does this cover all the various markdown content which a field name can have?

roanster007 replied:
it now covers mentions, silent mentions, bold/italic, url, <time> fields. emojis are anyways shown as `:smile:` so I guess yeah?

sahil839 reviewed (commented):
Posted few comments.

roanster007 commented:
conflicts are removed.

alya commented:
@sahil839 any more feedback from your side?

sahil839 commented on zerver/actions/custom_profile_fields.py:
```diff
@@ -108,7 +119,9 @@ def try_update_realm_custom_profile_field(
     field.name = name
+    field.rendered_name = render_custom_profile_fields(name, realm)
     field.hint = hint
+    field.rendered_hint = render_custom_profile_fields(hint, realm)
```
Would be worth adding a small test verifying the `rendered_...` fields on creating and updating custom profile fields.

roanster007 replied:
added!

sahil839 reviewed (commented):
Posted comments. I am not sure about what markdown we should allow like would some consider to use user/user group mention syntax in custom profile field label and hints. Not sure if there is any easy way to restrict to a subset of markdown syntax though.

roanster007 commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,11 @@ format used by the Zulip server that they are interacting with.
+**Feature level 248**
+
+* [`POST /register`](/api/register-queue): Added new fields
+  `rendered_name` and `rendered_hint`.
```
done!

roanster007 commented:
addressed the review comments:

> Would be worth adding a small test verifying the rendered_... fields on creating and updating custom profile fields.

added tests

> Should mention the objects in which these fields are added.

updated.

>I am not sure about what markdown we should allow like would some consider to use user/user group mention syntax in custom profile field label and hints. Not sure if there is any easy way to restrict to a subset of markdown syntax though.

I don't think there's some way to do that AFAIK since we always convert the markdown to html elements, but here we are rather trying to play around with the original data to display in tooltips, rather than the markdown rendered ones.

sahil839 commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,11 @@ format used by the Zulip server that they are interacting with.
 
+**Feature level 257**
+
+* [`POST /register`](/api/register-queue): Added new fields `rendered_name`
```
This should also mention other endpoints like `GET /events` and the `GET` endpoint for custom profile fields. And in the `POST /register` response as well, these fields are added in objects of `custom_profile_fields` object as well. You can have a look at existing changes in this file to get an idea of how to write these.

sahil839 commented on zerver/lib/event_schema.py:
```diff
@@ -174,6 +174,8 @@
     ],
     optional_keys=[
         ("display_in_profile_summary", bool),
+        ("rendered_name", str),
```
These fields are always included, so this should be in the `required_keys` with type set to `OptionalType(str)` to handle the case where it can be `None`.

sahil839 commented on web/templates/confirm_dialog/confirm_delete_profile_field.hbs:
```diff
@@ -1,12 +1,24 @@
+        {{#*inline "z-profile-field-name"}}
+        {{#if rendered_profile_field_name}}
+        <strong class="profile_field_name rendered_markdown">{{rendered_markdown rendered_profile_field_name}}</strong>
+        {{else}}
```
This if-conditions are required now, but am not sure we would need this if we add a migration to set the `rendered_..` fields for existing custom profile fields as well. We have these for stream descriptions as well, but maybe we can remove them as well.

There is no need of any changes for now till we finalize whether we want a migration for existing fields, but just mentioning it here so we do not miss this.

sahil839 commented on zerver/tests/test_custom_profile_data.py:
```diff
@@ -80,14 +80,18 @@ def test_create(self) -> None:
+        rendered_hint = '<p><span class="user-mention" data-user-id="11">@Iago</span></p>'
         self.assertEqual(field.id, field.order)
+        self.assertEqual(field.rendered_name, rendered_name)
+        self.assertEqual(field.rendered_hint, rendered_hint)
```
We should test this for updating fields as well.

sahil839 replied:
And for default external fields as well. Having these tests would be good to make sure that we do not accidentally remove those lines to set the `rendered_..` fields.

sahil839 reviewed (commented):
Posted few comments.

kuv2707 commented:
@roanster007 Should I take this up?

roanster007 commented:
@kuv2707 yes, sure!
